Data/IP/Addr.hs

```diff
@@ -258,85 +259,186 @@ showIPv6 ip@(IP6 (a1,a2,a3,a4))
 --
 
 {-|
-  The 'toIPv4' function takes a list of 'Int' and returns 'IPv4'.
+  The 'toIPv4' function returns the 'IPv4' address corresponding to the given
+  list of 'Int' octets.  The function is strict in the four elements of the
+  list.  An error is returned if the list has a differnet length.  The input
+  elements are silently truncated to their 8 least-significant bits before they
+  are combined to form the IPv4 address.
 
 >>> toIPv4 [192,0,2,1]
 192.0.2.1
 -}
 toIPv4 :: [Int] -> IPv4
-toIPv4 = IP4 . toWord32
+toIPv4 [a1, a2, a3, a4] = IP4 w
   where
-    toWord32 [a1,a2,a3,a4] = fromIntegral $ shift a1 24 + shift a2 16 + shift a3 8 + a4
-    toWord32 _             = error "toWord32"
+    !w = (fromIntegral a1 .&. 0xff) `unsafeShiftL` 24 .|.
```

Comment:
I believe that `BangPatterns` is not necessary because this biding is strict thanks to `Strict`.

Reply:
Yes, in this case the BangPattern is redundant. Even without `Strict` or `StrictData` because `IP4` is a newtype for `Word32`, so unavoidably strict. The `w` values is unavoidably evaluated whever the `IP4` is evaluated. The only time it could be needed would be if the definition changed from a `newtype` to `data IPv4Addr = IP4 Word32` and not `data IPv4Addr = IP4 ! Word32`, but that would be a mistake not to make.

Reply:
I dropped the redundant (though harmless) BangPattern. GHC should be able to figure out that this function is strict in both its arguments and the output Word32 (since newtype).